oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/authentication/OAuth2TokenExchangeAuthenticationProvider.java

```diff
@@ -0,0 +1,313 @@
+				throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_GRANT);
+			}
+		}
+
+		Set<String> authorizedScopes = Collections.emptySet();
+		if (!CollectionUtils.isEmpty(tokenExchangeAuthentication.getScopes())) {
+			for (String requestedScope : tokenExchangeAuthentication.getScopes()) {
+				if (!registeredClient.getScopes().contains(requestedScope)) {
+					throw new OAuth2AuthenticationException(OAuth2ErrorCodes.INVALID_SCOPE);
+				}
+			}
+			authorizedScopes = new LinkedHashSet<>(tokenExchangeAuthentication.getScopes());
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Validated token request parameters");
+		}
+
+		if (CollectionUtils.isEmpty(authorizedScopes)) {
+			// TODO: Validate that original authorizedScopes are also configured for this client?
+			authorizedScopes = new LinkedHashSet<>(subjectAuthorization.getAuthorizedScopes());
+		}
+
+		// @formatter:off
+		DefaultOAuth2TokenContext.Builder tokenContextBuilder = DefaultOAuth2TokenContext.builder()
+				.registeredClient(registeredClient)
+				.authorization(subjectAuthorization)
+				.principal(subjectPrincipal)
+				.authorizationServerContext(AuthorizationServerContextHolder.getContext())
+				.authorizedScopes(authorizedScopes)
+				.tokenType(OAuth2TokenType.ACCESS_TOKEN)
+				.authorizationGrantType(TOKEN_EXCHANGE)
+				.authorizationGrant(tokenExchangeAuthentication);
+		// @formatter:on
+		if (actorPrincipal != null) {
+			// NOTE: Passing attributes requires corresponding changes in JwtGenerator.
+			// TODO: Consider a "composite" principal combining the subject and actor principals?
+			tokenContextBuilder.put(Authentication.class.getName().concat(".ACTOR"), actorPrincipal);
+		}
+
+		// ----- Access token -----
+		OAuth2TokenContext tokenContext = tokenContextBuilder.build();
+		OAuth2Token generatedAccessToken = this.tokenGenerator.generate(tokenContext);
+		if (generatedAccessToken == null) {
+			OAuth2Error error = new OAuth2Error(OAuth2ErrorCodes.SERVER_ERROR,
+					"The token generator failed to generate the access token.", ERROR_URI);
+			throw new OAuth2AuthenticationException(error);
+		}
+
+		if (this.logger.isTraceEnabled()) {
+			this.logger.trace("Generated access token");
+		}
+
+		OAuth2AccessToken accessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER,
+				generatedAccessToken.getTokenValue(), generatedAccessToken.getIssuedAt(),
+				generatedAccessToken.getExpiresAt(), tokenContext.getAuthorizedScopes());
+
+		// @formatter:off
+		OAuth2Authorization.Builder authorizationBuilder = OAuth2Authorization.withRegisteredClient(registeredClient)
+				.principalName(subjectAuthorization.getPrincipalName())
+				.authorizationGrantType(TOKEN_EXCHANGE)
+				.authorizedScopes(authorizedScopes)
+				.attribute(SUBJECT_AUTHORIZATION_ID, subjectAuthorization.getId())
+				.attribute(SUBJECT_TOKEN_TYPE, tokenExchangeAuthentication.getSubjectTokenType())
+				.attribute(Principal.class.getName(), subjectPrincipal);
```

Comment:
I feel we need some way to store information about the original subject. I'm open to suggestions for a preferred way to store this information or if this ok for now. The principal is important though, as we need to carry that forward if this issued token is later exchanged for yet another token.

Reply:
I'm not sure if we should store each item in their own attribute. Maybe we create a simple "holder" class and it's composed of all the items, which is then stored in an attribute of the authorization.